dragging pieces to squares

cbohara opened the pull request:
I changed the game to have gray and white squares with a black border
I took it out of the booyah-box because it wasn't looking too good
Most importantly I got the pieces bigger and they fit in the center of the squares

I have created the ability to drag pieces to squares
They snap to the center of squares

I have no idea how to sync up the logic of the pieces to the movement
I can use your advice @jacobwgillespie !

jacobwgillespie commented on app/models/piece.rb:
```diff
@@ -124,8 +124,9 @@ def never_moved?
-  # moves piece to the destination square
-  def move_to!(x,y)
+  def move_to!(params)
+    x = params[:x_position].to_i
```
You should do this in the controller, not the model, because the model cares about x and y, and the controller cares about HTTP requests and responses (params).

jacobwgillespie replied:
And this is also what's causing all the test failures.

jacobwgillespie commented on app/views/games/_chessboard.html.erb:
```diff
@@ -1,42 +1,61 @@
-  })
- });
+$(document).ready(function () {
+  $(function() {
```
This syntax (`$(function() {`) is functionally identical to the line above it (`$(document).ready(function () {`) - they're both ways to set up a callback for when the page is loaded.

jacobwgillespie commented on app/views/games/_chessboard.html.erb:
```diff
@@ -1,42 +1,61 @@
+
+        $.ajax({
+          type: 'PATCH',
+          url: '/pieces/' + piece.id,
```
You will need to add an additional line here:

```javascript
beforeSend: $.rails.CSRFProtection,
```

Rails prevents cross-site request forging (CSRF) with a token.  A short description of CSRF is where a webpage tricks your browser into sending a request it did not intend to another server.  Like suppose you had www.badsite.com that had an image tag like so:

```html
<img src="https://www.facebook.com/user/delete-account/confirm" />
```

If that was the real delete-your-account URL and Facebook didn't have CSRF protection, that might cause anybody who visited that webpage to suddenly lose their FB account.

That's a simple example, and it can get more complex...

Rails solves this by generating a unique token for every page load and requiring that token to be present for any PUT / PATCH / POST / DELETE request.  That's what this line is outputting: https://github.com/team-rook/chess-app/blob/bc17f18c170d708a01e8adb35f6c9d10bf84c33b/app/views/layouts/application.html.erb#L7

The above line will cause jQuery to include that special token in your ajax request here - otherwise Rails will ignore it as a security measure.

jacobwgillespie commented on app/controllers/pieces_controller.rb:
```diff
@@ -0,0 +1,13 @@
+	def update
+    @piece = Piece.find(params[:id])
+    @game = @piece.game
+    @piece.move_to!(piece_params)
```
You should wrap this in an `if` statement to ensure that the piece.user_id is equal to the current_user.id so that players cannot move pieces that aren't theirs.

Also, at the end you should likely redirect the user back to the game, so something like `redirect_to @game` as the last line of the method.

jacobwgillespie replied:
Oh, wait, ignore that - this is an AJAX thing.  You probably want to respond with a success / fail message like so:

```ruby
# either
render json: 'success'

# or
render json: 'failure'
```

That way the JS on the frontend can let the user know if it worked.

jacobwgillespie commented on app/views/games/_chessboard.html.erb:
```diff
@@ -1,42 +1,60 @@
+          <td class="square" data-x-position="<%=column%>" data-y-position="<%=row%>">
+          <% piece = @game.piece_data[column][row]%>
+          <% if piece.present? %>
+            <div class="piece" data-piece-id="<%=piece%>">
```
You probably want `<%= piece.id %>`

cbohara commented on app/controllers/pieces_controller.rb:
```diff
@@ -0,0 +1,20 @@
+	def update
+	  @piece = Piece.find(params[:id])
+	  @game = @piece.game
+		if @piece.user_id == current_user.id
```
@jacobwgillespie @hathcode i'm not sure if this is correct

cbohara commented:
@jacobwgillespie I've tried adding some logic to the piece controller but I'm not sure if that's what I should be doing.  

My struggle is that I can't control who can move which piece, and I can't control how the pieces move :(

cbohara commented:
I realized we also need to remove captured pieces from the board.  Right now when the page refreshes, captured pieces are removed from the board.  Still trying to figure out a way to remove the piece without a page refresh.

cbohara commented:
I made some design updates for consistency and readability 
I also changed the font to spice things up a bit
I added a flash alert to let the white user know they are waiting for another player to join the game when they see a chess board with no pieces on it
And most importantly, I got the piece html removed from the board without a page refresh and without deleting the piece info from the database

kyidkim commented:
👍
